bevm/bevm_call_byzcoin_test.go

```diff
@@ -0,0 +1,514 @@
+package bevm
+
+import (
+	"math/big"
+	"testing"
+	"time"
+
+	"go.dedis.ch/cothority/v3"
+	"go.dedis.ch/cothority/v3/byzcoin"
+	"go.dedis.ch/cothority/v3/darc"
+	"go.dedis.ch/onet/v3"
+	"go.dedis.ch/onet/v3/log"
+	"golang.org/x/xerrors"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/stretchr/testify/require"
+)
+
+func init() {
+	err := byzcoin.RegisterGlobalContract(myValueContractID,
+		myValueContractFromBytes)
+	if err != nil {
+		log.ErrFatal(err)
+	}
+
+	evmSpawnableContracts[myValueContractID] = true
+}
+
+const myValueContractID = "MyValueContract"
+
+func myValueContractFromBytes(in []byte) (byzcoin.Contract, error) {
+	return myValueContract{value: in}, nil
+}
+
+// The test value contracts just holds a value
+type myValueContract struct {
+	byzcoin.BasicContract
+	value []byte
+}
+
+func (c myValueContract) Spawn(rst byzcoin.ReadOnlyStateTrie,
+	inst byzcoin.Instruction, cin []byzcoin.Coin) (sc []byzcoin.StateChange,
+	cout []byzcoin.Coin, err error) {
+	cout = cin
+
+	// Find the darcID for this instance.
+	var darcID darc.ID
+	_, _, _, darcID, err = rst.GetValues(inst.InstanceID.Slice())
+	if err != nil {
+		return
+	}
+
+	var newInstanceID byzcoin.InstanceID
+	seed := inst.Spawn.Args.Search("seed")
+	if seed != nil {
+		newInstanceID = byzcoin.ComputeNewInstanceID(myValueContractID, seed)
+	} else {
+		newInstanceID = byzcoin.NewInstanceID(inst.DeriveID("").Slice())
+	}
+
+	sc = []byzcoin.StateChange{
+		byzcoin.NewStateChange(
+			byzcoin.Create,
+			newInstanceID,
+			myValueContractID,
+			inst.Spawn.Args.Search("value"),
+			darcID),
+	}
+
+	return
+}
+
+func (c myValueContract) Invoke(rst byzcoin.ReadOnlyStateTrie,
+	inst byzcoin.Instruction, cin []byzcoin.Coin) (sc []byzcoin.StateChange,
+	cout []byzcoin.Coin, err error) {
+	cout = cin
+
+	// Find the darcID for this instance.
+	var darcID darc.ID
+
+	_, _, _, darcID, err = rst.GetValues(inst.InstanceID.Slice())
+	if err != nil {
+		return
+	}
+
+	switch inst.Invoke.Command {
+	case "update":
+		sc = []byzcoin.StateChange{
+			byzcoin.NewStateChange(
+				byzcoin.Update,
+				inst.InstanceID,
+				myValueContractID,
+				inst.Invoke.Args.Search("value"),
+				darcID),
+		}
+		return
+
+	default:
+		return nil, nil, xerrors.New("Value contract can only update")
+	}
+}
+
+func (c myValueContract) Delete(rst byzcoin.ReadOnlyStateTrie,
+	inst byzcoin.Instruction, cin []byzcoin.Coin) (sc []byzcoin.StateChange,
+	cout []byzcoin.Coin, err error) {
+	cout = cin
+
+	// Find the darcID for this instance.
+	var darcID darc.ID
+
+	_, _, _, darcID, err = rst.GetValues(inst.InstanceID.Slice())
+	if err != nil {
+		return
+	}
+
+	sc = []byzcoin.StateChange{
+		byzcoin.NewStateChange(
+			byzcoin.Remove,
+			inst.InstanceID,
+			myValueContractID,
+			nil,
+			darcID),
+	}
+	return
+}
+
+// Inspired by byzcoin.Client.WaitProof()
+func waitProofGone(t *testing.T, cl *byzcoin.Client, interval time.Duration,
```

Comment:
Can you try to use `cl.UseNode(1)` in your tests and then remove the `waitProofGone` and other `WaitProof`s?

Reply:
> Can you try to use `cl.UseNode(1)` in your tests and then remove the `waitProofGone` and other `WaitProof`s?

How do I then check that an InstanceID exists (or no longer does)? Directly with `GetProof()`/`InclusionProof.Exists()`?

Reply:
yes, that should work. The problem that `waitProofGone` and `WaitProof` solve is the communication with different nodes: if you send your transaction to node A, and get the proof from node B, you're not sure they are synchronized.

`cl.UseNode(1)` puts all communication through the same node. So if you send a transaction and it returns, you can be sure that the block has been applied in node 1.